ingestion/src/lib.rs

```diff
@@ -0,0 +1,242 @@
+                    DO UPDATE SET obsvalue = EXCLUDED.obsvalue",
+        )
+        .await?;
+
+    let mut futures = data
+        .iter()
+        .map(|datum| async {
+            conn.execute(
+                &query,
+                &[&datum.timeseries_id, &datum.timestamp, &datum.value],
+            )
+            .await
+        })
+        .collect::<FuturesUnordered<_>>();
+
+    while let Some(res) = futures.next().await {
+        res?;
+    }
+
+    Ok(())
+}
+
+pub mod kldata;
+use kldata::{filter_and_label_kldata, parse_kldata};
+
+/// Format of response Obsinn expects from this API
+#[derive(Debug, Serialize)]
+struct KldataResp {
+    /// Optional message indicating what happened to the data
+    message: String,
+    /// Should be the same message_id we received in the request
+    message_id: usize,
+    /// Result indicator, 0 means success, anything else means fail.
+    // Kvalobs uses some specific numbers to denote specific errors with this, I don't much see
+    // the point, the only information Obsinn can really action on as far as I can tell, is whether
+    // we failed and whether it can retry
+    res: u8, // TODO: Should be an enum?
+    /// Indicates whether Obsinn should try to send the message again
+    retry: bool,
+}
+
+async fn handle_kldata(
+    State(pool): State<PgConnectionPool>,
+    State(param_conversions): State<ParamConversions>,
+    State(permit_table): State<Arc<RwLock<(ParamPermitTable, StationPermitTable)>>>,
+    body: String,
+) -> Json<KldataResp> {
+    let result: Result<usize, Error> = async {
+        let mut conn = pool.get().await?;
+
+        let (message_id, obsinn_chunk) = parse_kldata(&body)?;
+
+        let data =
+            filter_and_label_kldata(obsinn_chunk, &mut conn, param_conversions, permit_table)
+                .await?;
+
+        insert_data(data, &mut conn).await?;
+
+        Ok(message_id)
+    }
+    .await;
+
+    match result {
+        Ok(message_id) => Json(KldataResp {
+            message: "".into(),
+            message_id,
+            res: 0,
+            retry: false,
+        }),
+        Err(e) => Json(KldataResp {
+            message: e.to_string(),
+            message_id: 0, // TODO: some clever way to get the message id still if possible?
+            res: 1,
+            retry: !matches!(e, Error::Parse(_)),
+        }),
+    }
+}
+
+pub async fn run(
+    connect_string: &str,
+    param_conversion_path: &str,
+    permit_tables: Arc<RwLock<(ParamPermitTable, StationPermitTable)>>,
+) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
+    // set up postgres connection pool
+    let manager = PostgresConnectionManager::new_from_stringlike(connect_string, NoTls)?;
+    let db_pool = bb8::Pool::builder().build(manager).await?;
+
+    // set up param conversion map
+    // TODO: extract to separate function?
+    let param_conversions = Arc::new(
+        csv::Reader::from_path(param_conversion_path)
+            .unwrap()
```

Comment:
possible panic that needs to be handled? 
for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#missing_panics_doc

Reply:
I think this is fine, this will only panic if the path is wrong/missing, and it will panic at startup.

Did you get this lint from running clippy locally though? Wondering why it doesn't show in CI. If it really is linting though then we should add a doc or annotation to silence it

Reply:
I ran clippy locally in pedantic mode, and kinda scanned through for stuff that I thought was potentially useful or maybe dangerous. I was wrong about this one being relevant, sorry!

Reply:
I'm not sure how helpful that is tbh, the things that are gated into pedantic are there because they aren't considered an actual problem, they're nitpicks even by my standards.